File: cv-parser/choix_modele.py
"""Choix du modèle d'extraction d'un CV, par action (#194).

Le catalogue partagé (document-parsing/models/catalogue.json et son chargeur
catalogue.py, source unique) dit quels modèles proposer pour la tâche
`resume`, sur quelle voie DocIE, avec quel identifiant réel (variables
DOCIE_MODELE_<MODELE> et DOCIE_AGENT_RESUME_<MODELE>). Ce module ne fait que
l'appliquer à cv-parser :

- la VOIE d'un fichier suit la configuration existante, sans rien réaiguiller :
  bridge actif (DOCIE_EXTRACTION_ENABLED) et PDF -> voie `agent`
  (docie_bridge_extraction) ; client historique en mode inline -> voie
  `texte` (docie_client, PDF texte ou DOCX) ; mode studio -> aucune voie du
  catalogue, donc aucun sélecteur ;
- le sélecteur ne s'affiche que si PLUSIEURS modèles sont proposés : tant
  qu'aucune alternative n'est configurée, l'écran ne change pas ;
- un choix EXPLICITE = un champ `modele` non vide reçu avec le dépôt ou la
  ré-analyse. Le sélecteur, quand il est affiché, l'envoie toujours (défaut
  compris) : l'utilisateur a vu le modèle nommé à côté du bouton. Sans champ,
  comportement d'avant le catalogue, repli compris ;
- un choix est vérifié sur le document RÉEL juste avant l'envoi (lignes non
  vides pour la voie texte, pages pour la voie agent) et n'est jamais
  remplacé par un autre modèle : refus nommé (`modele_non_propose`,
  `limite`, `configuration`), levé comme ErreurTache — la tâche échoue avec
  ce code, sans fiche vide de repli.

Le chargeur est lu par son chemin (importlib), jamais par sys.path : dans
l'image, il est copié sous /document-parsing/models/ (voir le Dockerfile),
même profondeur relative que dans le dépôt.
"""
import importlib.util
import os
from pathlib import Path

from docie_bridge_extraction import docie_extraction_enabled
from taches_upload import ErreurTache

TACHE = "resume"
CHEMIN_CHARGEUR = Path(__file__).resolve().parents[1] / "document-parsing" / "models" / "catalogue.py"

# Extensions que le bridge reçoit en fichier (docie_bridge_extraction._MIME_BY_SUFFIX).
_EXT_BRIDGE = (".pdf", ".png", ".jpg", ".jpeg", ".webp")
# Extensions que le client historique lit en texte en mode inline (docie_client.extract_resume).
_EXT_TEXTE = (".pdf", ".docx")

_catalogue = None


def charger():
    """Le module catalogue.py, chargé une fois. Absent : FileNotFoundError."""
    global _catalogue
    if _catalogue is None:
        spec = importlib.util.spec_from_file_location("adbi_catalogue_modeles", CHEMIN_CHARGEUR)
        module = importlib.util.module_from_spec(spec)
        spec.loader.exec_module(module)
        _catalogue = module
    return _catalogue


def voie_pour(ext):
    """Voie DocIE que prendra un fichier de cette extension, ou None."""
    ext = str(ext or "").lower()
    if docie_extraction_enabled() and ext in _EXT_BRIDGE:
        return "agent"
    if os.environ.get("DOCIE_EXTRACTION_MODE", "inline") == "inline" and ext in _EXT_TEXTE:
        return "texte"
    return None


def modeles_proposes(ext=None):
    """Modèles du sélecteur, défaut d'abord : [{id, libelle, description, role}].

    `ext` connu (ré-analyse) : la voie de ce fichier. Absent (dépôt, PDF ou
    DOCX, en lot) : les voies des deux formats réunies ; un modèle configuré
    sur une seule voie est vérifié sur le fichier réel au moment de l'envoi.
    Les limites ne sont pas évaluées ici (document inconnu).
    """
    voies = []
    for e in ([ext] if ext else [".pdf", ".docx"]):
        voie = voie_pour(e)
        if voie and voie not in voies:
            voies.append(voie)
    if not voies:
        return []
    catalogue = charger()
    vus = {}
    for voie in voies:
        for offre in catalogue.modeles_offerts(TACHE, voie):
            vus.setdefault(offre["id"], {"id": offre["id"], "libelle": offre["libelle"],
                                         "description": offre["description"], "role": offre["role"]})
    return sorted(vus.values(), key=lambda o: o["role"] != "defaut")


def offres_par_format():
    """{".pdf": [ids], ".docx": [ids]} : modèles proposés pour la voie de chaque
    format accepté au dépôt. Le navigateur envoie `modele` pour un fichier dès
    qu'au moins un modèle est proposé pour SON format (même règle que contrats,
    #210), sélecteur affiché ou non."""
    catalogue = None
    offres = {}
    for ext in (".pdf", ".docx"):
        voie = voie_pour(ext)
        if voie is None:
            offres[ext] = []
            continue
        catalogue = catalogue or charger()
        offres[ext] = [o["id"] for o in catalogue.modeles_offerts(TACHE, voie)]
    return offres


class Choix:
    """Un modèle explicitement choisi, vérifié voie par voie sur le document réel.

    Passé à docie_client.extract_resume (`pour_texte`) et à
    docie_bridge_extraction.extract_resume (`pour_agent`), qui l'appellent
    juste avant l'appel réseau et envoient l'identifiant rendu.
    """

    def __init__(self, modele):
        self.modele = modele
        self.offre = None

    def _choisir(self, voie, document):
        catalogue = charger()
        try:
            self.offre = catalogue.choisir_modele(TACHE, voie, self.modele, document=document)
        except catalogue.CatalogueError as exc:
            raise ErreurTache(str(exc), code=exc.code) from None
        return self.offre["identifiant"]

    def verifier(self, ext):
        """Avant tout travail : le modèle est-il proposé pour la voie de ce fichier ?"""
        voie = voie_pour(ext)
        if voie is None:
            # Même message que le catalogue pour un modèle non configuré.
            return self._choisir("__aucune__", None)
        return self._choisir(voie, None)

    def pour_texte(self, texte):
        return self._choisir("texte", {"lignes_non_vides": charger().compter_lignes_non_vides(texte)})

    def pour_agent(self, pages):
        """`pages` None : nombre de pages illisible. Refusé si le modèle a une
        limite de pages — une limite non vérifiable n'est pas une limite respectée."""
        if pages is None:
            offre = self._choisir("agent", None)
            if self.offre["limites"].get("pages_max") is not None:
                raise ErreurTache(
                    f"{self.offre['libelle']} : nombre de pages du document illisible, "
                    f"limite de {self.offre['limites']['pages_max']} pages non vérifiable.", code="limite")
            return offre
        return self._choisir("agent", {"pages": pages})


def resultat_partiel(metadata, data):
    """Résultat partiel d'une extraction (#203) : [{champ, raison}], ou None si
    on n'a pas pu le vérifier.

    Voie agent : relevé par le bridge (`metadata["partiel"]`). Voie texte (client
    historique, hors bridge) : même relevé, par la fonction du bridge — une
    seule reconnaissance des libellés DocIE, jamais recopiée ici.
    """
    partiel = (metadata or {}).get("partiel")
    if isinstance(partiel, list):
        return partiel
    try:
        from docie_bridge_extraction import _load_bridge
        return _load_bridge().resultat_partiel((metadata or {}).get("validation"), data)
    except Exception:
        return None


def modele_servi(voie, metadata):
    """Le modèle qui a RÉELLEMENT servi : {id, libelle} ou None.

    Lu dans la réponse (`model` voie texte, `agent` voie agent), rapproché des
    identifiants configurés avec ou sans `store:` ; sans correspondance, le nom
    brut rapporté par DocIE, jamais le modèle demandé. Même forme que contrats
    (#210) : l'identifiant `store:` configuré n'est pas recopié sur la fiche.
    Catalogue illisible : le nom brut, jamais une exception.
    """
    try:
        servi = charger().modele_servi(TACHE, voie, metadata)
    except Exception:
        brut = (metadata or {}).get("agent" if voie == "agent" else "model")
        servi = {"id": None, "libelle": brut} if isinstance(brut, str) and brut.strip() else None
    return {"id": servi["id"], "libelle": servi["libelle"]} if servi else None

File: one-pager/lib/choix-modele.js
"use strict";

/**
 * Choix du modele d'extraction d'un CV importe, par action (#194).
 *
 * Le catalogue partage (document-parsing/models/catalogue.json et son chargeur
 * catalogue.js, source unique) dit quels modeles proposer pour la tache
 * `resume`, sur quelle voie DocIE, avec quel identifiant reel
 * (DOCIE_MODELE_<MODELE> voie texte, DOCIE_AGENT_RESUME_<MODELE> voie agent).
 * Ce module ne fait que l'appliquer a one-pager :
 *
 * - la VOIE suit l'aiguillage existant de lib/import-pipeline (#180) : PDF ->
 *   voie agent (fichier), depot texte et .docx -> voie texte ;
 * - rien n'est propose tant que DOCIE_EXTRACTION_ENABLED n'est pas actif :
 *   proposer un modele que l'import ne peut pas appeler finirait en analyse
 *   locale, precisement ce que le choix interdit ;
 * - le selecteur (public/app.js) ne s'affiche qu'a partir de DEUX modeles ;
 * - un choix EXPLICITE = un champ `modele` non vide dans POST /api/import. Le
 *   selecteur, quand il est affiche, l'envoie toujours (defaut compris). Il est
 *   verifie sur le document REEL avant l'envoi (lignes non vides, pages) et
 *   n'est jamais remplace : refus nomme (CatalogueError `modele_non_propose`,
 *   `limite`, `configuration`), jamais de repli local.
 *
 * Le chargeur n'est `require` qu'a l'appel, comme le bridge : sans le fichier
 * (image sans la copie du Dockerfile), l'import sans choix fonctionne encore.
 */

const TACHE = "resume";

/*
 * Modeles HORS ADBI (#194) : le catalogue ne les propose que si le consommateur
 * les demande (option `externes`) ET que la cle du fournisseur est renseignee.
 * Voie TEXTE uniquement — le fournisseur ne declare que `texte`
 * (catalogue.json), et la voie agent enverrait le document LUI-MEME, pas son
 * texte, donc un scan entier plutot que ce que nous en avons lu.
 *
 * DONNEES PERSONNELLES : un CV est la donnee d'un CANDIDAT, pas un document
 * d'entreprise comme les cinq pieces de contrats. #194 avait ecarte le CV pour
 * cette raison (« en attente du proprietaire ») ; le proprietaire a tranche, a
 * la condition que l'utilisateur soit AVERTI, au moment du choix, que le texte
 * part chez un tiers — voir public/app.js. Jamais un defaut, jamais un repli.
 */
const EXTERNES = true;

function chargerCatalogue() {
  // eslint-disable-next-line global-require
  return require("../../document-parsing/models/catalogue");
}

function chargerPont() {
  // eslint-disable-next-line global-require
  return require("../../document-parsing/bridge/docie-bridge");
}

// Noms des modeles prets sur le store DocIE, dernier releve du pont ; pont absent : [].
function storePret() {
  try { return chargerPont().storeUtilisableConnu().map((m) => m.nom).filter(Boolean); } catch { return []; }
}

// Relit le store (cache 5 min du pont) avant un rendu de selecteur ou une verification.
async function rafraichirStore(env = process.env) {
  try { await chargerPont().storeUtilisable({ env }); } catch { /* releve precedent conserve */ }
}

function docieActif(env) {
  // eslint-disable-next-line global-require
  return require("./import-pipeline").docieActif(env);
}

/**
 * { texte: [ids], agent: [ids] } : modeles proposes par voie. Le navigateur
 * envoie `modele` pour un fichier des qu'au moins un modele est propose pour SA
 * voie (PDF -> agent, texte et .docx -> texte), selecteur affiche ou non — meme
 * regle que contrats (#210).
 */
function offresParVoie(env = process.env) {
  if (!docieActif(env)) return { texte: [], agent: [] };
  const catalogue = chargerCatalogue();
  const store = storePret();
  const offres = {};
  for (const voie of ["texte", "agent"]) {
    offres[voie] = catalogue.modelesOfferts(TACHE, voie, { env, externes: voie === "texte" && EXTERNES, store }).map((o) => o.id);
  }
  return offres;
}

/** Modeles du selecteur, defaut d'abord : [{ id, libelle, description, role, experimental }]. */
function modelesProposes(env = process.env) {
  if (!docieActif(env)) return [];
  const catalogue = chargerCatalogue();
  const store = storePret();
  const vus = new Map();
  // Voies des formats acceptes, reunies : le format du prochain fichier n'est
  // pas connu ; un modele configure sur une seule voie est verifie a l'envoi.
  for (const voie of ["texte", "agent"]) {
    for (const o of catalogue.modelesOfferts(TACHE, voie, { env, externes: voie === "texte" && EXTERNES, store })) {
      if (!vus.has(o.id)) {
        vus.set(o.id, { id: o.id, libelle: o.libelle, description: o.description, role: o.role, experimental: o.experimental === true });
      }
    }
  }
  return [...vus.values()].sort((a, b) => (a.role === "defaut" ? 0 : 1) - (b.role === "defaut" ? 0 : 1));
}

/**
 * Le modele `modele` pour `voie`, verifie sur `document`
 * ({ lignesNonVides } voie texte, { pages } voie agent, pages null = illisible).
 * Leve CatalogueError ; ne rend jamais un autre modele que celui demande.
 */
function choisir(voie, modele, document, env = process.env) {
  const catalogue = chargerCatalogue();
  const illisible = voie === "agent" && document.pages == null;
  const offre = catalogue.choisirModele(TACHE, voie, {
    env, modele, document: illisible ? null : document, externes: voie === "texte" && EXTERNES, store: storePret(),
  });
  if (illisible && typeof offre.limites.pages_max === "number") {
    // Une limite non verifiable n'est pas une limite respectee.
    throw new catalogue.CatalogueError("limite",
      `${offre.libelle} : nombre de pages du document illisible, limite de ${offre.limites.pages_max} pages non vérifiable.`);
  }
  return offre;
}

function compterLignesNonVides(texte) {
  return chargerCatalogue().compterLignesNonVides(texte);
}

/** Pages d'un PDF (meme lecteur que lib/ingest), ou null s'il est illisible. */
async function compterPages(buffer) {
  try {
    const pdfjs = await import("pdfjs-dist/legacy/build/pdf.mjs");
    const doc = await pdfjs.getDocument({ data: new Uint8Array(buffer), isEvalSupported: false, useSystemFonts: true }).promise;
    const pages = doc.numPages;
    await doc.destroy();
    return pages;
  } catch {
    return null;
  }
}

/**
 * Modele qui a REELLEMENT servi, lu dans les metadonnees DocIE (`model` voie
 * texte, `agent` voie agent) : { id, libelle } ou null — meme forme que contrats
 * (#210), sans l'identifiant configure. Sans correspondance, le nom brut rapporte
 * par DocIE, jamais le modele demande. Catalogue illisible : le nom brut, jamais
 * une exception.
 */
function modeleServi(voie, metadata, env = process.env) {
  let servi;
  try {
    servi = chargerCatalogue().modeleServi(TACHE, voie, { env, metadata, store: storePret() });
  } catch {
    const brut = voie === "agent" ? (metadata || {}).agent : (metadata || {}).model;
    servi = typeof brut === "string" && brut.trim() ? { id: null, libelle: brut } : null;
  }
  return servi ? { id: servi.id, libelle: servi.libelle } : null;
}

/**
 * Le modele choisi est-il servi par un fournisseur HORS ADBI ?
 *
 * Quand c'est vrai, `offre.identifiant` est le MODE du transport
 * (`rapide` / `raisonnement`) et non un profil DocIE : l'envoyer a
 * extraireTexteViaDocie demanderait a DocIE un modele nomme « rapide ».
 */
function estExterne(offre) {
  return Boolean(offre && typeof offre.fournisseur === "string" && offre.fournisseur);
}

module.exports = { modelesProposes, offresParVoie, choisir, estExterne, compterLignesNonVides, compterPages, modeleServi, chargerCatalogue, rafraichirStore, storePret, TACHE };
